runtime/include/runtime.h

```diff
@@ -8,6 +8,9 @@
 #include "likely.h"
 #include "types.h"
 
+/* Memory barrier to prevent unexpected code migration due to compiler optimization */
+#define barrier() __asm__ __volatile__("" ::: "memory")
```

Comment:
I don't get it. Unexpected code migration from what to what?

Reply:
Sorry for the delay here! Basically, an optimizing compiler will reorder instructions however it sees fit. If there is a variable protecting a critical section, this means that code may move out of or into the critical section, which could cause bugs. Inline assembly acts as a barrier that no assembly instructions can be reordered across. For us, we use this to ensure that code is either intentionally preemptable or non-preemptable.